types/src/transaction/mod.rs

```diff
@@ -631,19 +632,20 @@ impl SignedTransaction {
 
 #[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
 #[cfg_attr(any(test, feature = "fuzzing"), derive(Arbitrary))]
-pub struct TransactionWithProof {
+#[serde(bound = "for<'a> PS: Deserialize<'a>")]
+pub struct TransactionWithProof<PS: ProtocolSpec> {
```

Comment:
This is one thing that we should try to avoid: putting trait bounds on the definitions of structs. It results in a lot of extra stuff (like the serde bound you now need to add here). Instead we should try to structure these generics to only have bounds on impl blocks (of course this may not be possible in all cases). Eg here maybe we can instead have the generic parameter be of type `TransactionInfoTrait` directly instead of `PS::TransactionInfo`

Reply:
But 1. it sorta defeats the purpose of the "<PS> everywhere convenience" and 2. These structs goes through the wire and needs to have concrete types for their fields (same reason why they need to be SerDe?)

Reply:
I think pipping through the ProtocolSpec trait everywhere will probably lead to more headache because of places like this. The `TransactionInfoWithProof` type doesn't need to know about anything more than what the `TransactionInfo` trait supplies. I just think we can probably simplify things just a tinge to make them easier to work with.

Reply:
Think of the user of this struct, for example, the DB.

You either have DiemDB<PS>, in which case you'll write a lot of `TransactionListWithProof<PS::TransactionInfo>`, `AccountWithProof<PS::AccountInfo>` etc instead of `TransactionListWithProof<PS>` and `AccountWithProof<PS>`. 

Or you need to do `DiemDB<TxnInfo: TransactionInfoTrait, AccInfo: AccountInfoTrait, ...>`  which defeats the purpose of the `ProtocolSpec` trait.

I'd also argue the `#[serde(bound=)]` directive is the one time burden of the author of each core data structure authors, other developers won't notice. -- For example, I didn't find similar annoyances while I did the follow up commits around parameterizing the executor on PS.

Reply:
One hang up I have with the bound statement is that is not really accurate. Here we really care about TransactionInfo being `Deserialize` and not `PS`, since its not whats actually being serialized.

Reply:
Yes, I think that's really a unfortunate limitation of Rust. Or there's more magic for this purpose that I don't know yet..